tests/diagnostics/inout-never-written.slang

```diff
@@ -0,0 +1,47 @@
+//TEST:SIMPLE(filecheck=CHK): -target spirv
+
+struct State
+{
+    float3 v;
+    float3 n;
+    int rnd;
+};
+
+//CHK-DAG: ([[# @LINE + 1]]): warning 41022: inout parameter 'x' is never written to
+void int_never_assigned(inout int x) {}
+
+//CHK-DAG: ([[# @LINE + 1]]): warning 41022: inout parameter 'state' is never written to
+void state_never_assigned(inout State state, inout float v)
+{
+    v = state.v.x;
+}
+
+void state_assigned(inout State state)
+{
+    int3 vi = __slang_noop_cast <int3, float3> (state.v);
+    int3 ni = __slang_noop_cast <int3, float3> (state.n);
+    state.rnd = dot(vi, ni);
+}
+
+struct A
+{
+    int state;
+
+    //CHK-DAG: ([[# @LINE + 1]]): warning 41023: method marked `[mutable]` but never modifies `this`
+    [mutating] int next() { return state; }
+    
+    // This is ok because of the attribute
+    [semanticmutating] int progress() {}
+};
+
+__generic <T>
+struct B
+{
+    int state;
+
+    //CHK-DAG: ([[# @LINE + 1]]): warning 41023: method marked `[mutable]` but never modifies `this`
+    [mutating] int next() { return state; }
+};
```

Comment:
A case I am worried about is when the inheritance is involved.
Consider the following code snippet where one of derived structs mutates and the other one doesn't.
```
Interface Base
{
    [mutating] void doSomething();
};

struct NotMutating : Base
{
    [mutating] void doSomething() {} // warning will be printed.
};

struct Mutating : Base
{
    int a = 0;
    [mutating] void doSomething() { a = 1; }
}
```
Note that `[mutating]` is a part of the function signature in Slang.
Would your new attribute solve this problem if I do the following?
```
struct NotMutating : Base
{
    [semanticmutating] void doSomething() {} // is [mutating] still a part of this function signature?
};
```

Reply:
The idea is that it will indeed solve this case (I will confirm locally). The difference between `[semanticmutating]` and `[mutating]` is that the latter will force the `this` parameter to be `InOut` whereas this is not the case with `[semanticmutating]`.